jkube-kit/generator/karaf/src/test/java/org/eclipse/jkube/generator/karaf/KarafGeneratorTest.java

```diff
@@ -22,34 +24,166 @@
+  @DisplayName("No warnings or errors if the final output artifact is present")
+  void noWarnOrErrorIfFinalOutPutArtifactPresent() throws IOException {
+    // @Todo: move these initializations to setup method when replacing mocks with real objects
+    KitLogger testLogger = spy(new KitLogger.SilentLogger());
+    targetDir = Files.createDirectory(temporaryFolder.toPath().resolve("target")).toFile();
+    project = JavaProject.builder()
+            .baseDirectory(targetDir)
+            .buildDirectory(targetDir.getAbsoluteFile())
+            .buildPackageDirectory(targetDir.getAbsoluteFile())
+            .outputDirectory(targetDir)
+            .buildFinalName("sample")
+            .packaging("jar")
+            .build();
+    generatorContext = GeneratorContext.builder()
+            .project(project)
+            .logger(testLogger)
+            .build();
+
+    // Given
+    Files.createFile(targetDir.toPath().resolve("sample.jar"));
+    KarafGenerator testGenerator = new KarafGenerator(generatorContext);
+
+    // When
+    testGenerator.createDefaultAssembly();
+
+    // Then
+    verify(generatorContext.getLogger(), times(0)).warn(anyString());
+    verify(generatorContext.getLogger(), times(0)).error(anyString());
+  }
+
+  @Test
+  @DisplayName("Warn if the final output artifact is same as previous")
+  void warnIfFinalArtifactSameAsPrevious() throws IOException {
+    // @Todo: move these initializations to setup method when replacing mocks with real objects
+    KitLogger testLogger = spy(new KitLogger.SilentLogger());
+    targetDir = Files.createDirectory(temporaryFolder.toPath().resolve("target")).toFile();
+    project = JavaProject.builder()
+            .baseDirectory(targetDir)
+            .buildDirectory(targetDir.getAbsoluteFile())
+            .buildPackageDirectory(targetDir.getAbsoluteFile())
+            .outputDirectory(targetDir)
+            .buildFinalName("sample")
+            .packaging("jar")
+            .build();
+    generatorContext = GeneratorContext.builder()
+            .project(project)
+            .logger(testLogger)
+            .build();
+
+    // Given
+    Files.createFile(targetDir.toPath().resolve("sample.jar"));
+    KarafGenerator generator = new KarafGenerator(generatorContext);
+
+    // When
+    generator.createDefaultAssembly();
+    generator.createDefaultAssembly(); // calling twice to simulate that the final output artifacts were not changed
+
+    // Then
+    verify(generatorContext.getLogger())
+            .warn("karaf: Final output artifact is the same as previous build. " +
+                  "You might have forgotten to compile and package your application after making changes.");
+  }
+
+  @Test
+  @DisplayName("No warn or error if the final output artifact is present and different from previous")
+  void noWarnOrErrorIfFinalOutputNotSameAsPrevious() throws IOException {
+    // @Todo: move these initializations to setup method when replacing mocks with real objects
```

Comment:
I think addressing these TODO comments shouldn't be that hard. Please move common blocks to `@BeforeEach` methods.

Reply:
but that may conflict with other pre-existing test cases

Reply:
If it's conflicting, you can move these tests to a separate test class